apps/volunteer/choose_roles.py

```diff
@@ -108,3 +120,62 @@ def role(role_id):
         role=role,
         current_volunteer=current_volunteer,
     )
+
+
+@decorator
+def role_admin_required(f, *args, **kwargs):
+    if current_user.is_authenticated:
+        if int(args[0]) in [
+            ra.role_id for ra in current_user.volunteer_admin_roles
+        ] or current_user.has_permission("volunteer:admin"):
+            return f(*args, **kwargs)
+        abort(404)
+    return app.login_manager.unauthorized()
+
+
+@volunteer.route("role/<role_id>/admin")
+@role_admin_required
+def role_admin(role_id):
+    role = Role.query.get_or_404(role_id)
+    cutoff = datetime.now() - timedelta(minutes=30)
+    shifts = (
+        Shift.query.filter_by(role=role)
+        .filter(Shift.end >= cutoff)
+        .order_by(Shift.end)
+        .limit(5)
+        .all()
+    )
+    return render_template("volunteer/role_admin.html", role=role, shifts=shifts)
+
+
+@volunteer.route("role/<role_id>/toggle_arrived/<shift_id>/<user_id>")
```

Comment:
[can be merged but we should fix later] I think this should be a POST not a get as it's changing state. Not sure what @russss & @marksteward's thoughts on this are though so happy to be over ruled and can probably merge despite it.